File: sdks/go/types/schedule_at.go
package types

import (
	"github.com/clockworklabs/SpacetimeDB/sdks/go/bsatn"
)

// ScheduleAt is a sum type representing either a recurring interval or a specific time.
//   - Tag 0: Interval (TimeDuration)
//   - Tag 1: Time (Timestamp)
type ScheduleAt interface {
	bsatn.Serializable
	isScheduleAt()
}

// ScheduleAtInterval is the Interval variant of ScheduleAt (tag 0).
type ScheduleAtInterval struct {
	Value TimeDuration
}

func (ScheduleAtInterval) isScheduleAt() {}

func (s ScheduleAtInterval) WriteBsatn(w bsatn.Writer) {
	w.PutSumTag(0)
	s.Value.WriteBsatn(w)
}

// ScheduleAtTime is the Time variant of ScheduleAt (tag 1).
type ScheduleAtTime struct {
	Value Timestamp
}

func (ScheduleAtTime) isScheduleAt() {}

func (s ScheduleAtTime) WriteBsatn(w bsatn.Writer) {
	w.PutSumTag(1)
	s.Value.WriteBsatn(w)
}

// ReadScheduleAt reads a ScheduleAt from a BSATN reader.
func ReadScheduleAt(r bsatn.Reader) (ScheduleAt, error) {
	tag, err := r.GetSumTag()
	if err != nil {
		return nil, err
	}
	switch tag {
	case 0:
		v, err := ReadTimeDuration(r)
		if err != nil {
			return nil, err
		}
		return ScheduleAtInterval{Value: v}, nil
	case 1:
		v, err := ReadTimestamp(r)
		if err != nil {
			return nil, err
		}
		return ScheduleAtTime{Value: v}, nil
	default:
		return nil, &bsatn.ErrInvalidTag{Tag: tag, SumName: "ScheduleAt"}
	}
}

File: sdks/kotlin/src/commonMain/kotlin/com/clockworklabs/spacetimedb/ScheduleAt.kt
package com.clockworklabs.spacetimedb

import com.clockworklabs.spacetimedb.bsatn.BsatnReader
import com.clockworklabs.spacetimedb.bsatn.BsatnWriter
import com.clockworklabs.spacetimedb.protocol.TimeDuration

sealed class ScheduleAt {
    data class Interval(val every: TimeDuration) : ScheduleAt()
    data class Time(val time: Timestamp) : ScheduleAt()

    companion object {
        fun read(reader: BsatnReader): ScheduleAt {
            return when (reader.readTag().toInt()) {
                0 -> Interval(TimeDuration.read(reader))
                1 -> Time(Timestamp.read(reader))
                else -> throw IllegalStateException("Unknown ScheduleAt tag")
            }
        }

        fun write(writer: BsatnWriter, value: ScheduleAt) {
            when (value) {
                is Interval -> {
                    writer.writeTag(0u)
                    TimeDuration.write(writer, value.every)
                }
                is Time -> {
                    writer.writeTag(1u)
                    Timestamp.write(writer, value.time)
                }
            }
        }
    }
}
